packages/legend-query/src/components/QueryBuilderCustomDatePicker.tsx

```diff
@@ -0,0 +1,1267 @@
+    value: string,
+    duration: number,
+    unit: QUERY_BUILDER_CUSTOM_DATE_OPTION_UNIT | undefined,
+    direction: QUERY_BUILDER_CUSTOM_DATE_OPTION_DIRECTION | undefined,
+    referenceMoment:
+      | QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT
+      | undefined,
+  ) {
+    super(label, value);
+    this.duration = duration;
+    this.unit = unit;
+    this.direction = direction;
+    this.referenceMoment = referenceMoment;
+  }
+
+  generateDisplayableLabel(): string {
+    return [
+      this.duration,
+      this.unit,
+      this.direction,
+      this.referenceMoment,
+    ].join(' ');
+  }
+
+  updateLabel(): void {
+    this.label = this.generateDisplayableLabel();
+  }
+}
+
+class CustomFirstDayOfOption extends DatePickerOption {
+  unit: QUERY_BUILDER_FIRST_DAY_OF_UNIT | undefined;
+
+  constructor(
+    label: string,
+    value: string,
+    unit: QUERY_BUILDER_FIRST_DAY_OF_UNIT | undefined,
+  ) {
+    super(label, value);
+    this.unit = unit;
+  }
+}
+
+class CustomPreviousDayOfWeekOption extends DatePickerOption {
+  day: QUERY_BUILDER_DAY_OF_WEEK;
+
+  constructor(label: string, value: string, day: QUERY_BUILDER_DAY_OF_WEEK) {
+    super(label, value);
+    this.day = day;
+  }
+}
+
+const reservedCustomDateOptions: CustomDateOption[] = [
+  new CustomDateOption(
+    'Yesterday',
+    QUERY_BUILDER_CUSTOM_DATE_PICKER_OPTION.YESTERDAY,
+    1,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_UNIT.DAYS,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_DIRECTION.BEFORE,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT.TODAY,
+  ),
+  new CustomDateOption(
+    'One Week Ago',
+    QUERY_BUILDER_CUSTOM_DATE_PICKER_OPTION.ONE_WEEK_AGO,
+    1,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_UNIT.WEEKS,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_DIRECTION.BEFORE,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT.TODAY,
+  ),
+  new CustomDateOption(
+    'One Month Ago',
+    QUERY_BUILDER_CUSTOM_DATE_PICKER_OPTION.ONE_MONTH_AGO,
+    1,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_UNIT.MONTHS,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_DIRECTION.BEFORE,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT.TODAY,
+  ),
+  new CustomDateOption(
+    'One Year Ago',
+    QUERY_BUILDER_CUSTOM_DATE_PICKER_OPTION.ONE_YEAR_AGO,
+    1,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_UNIT.YEARS,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_DIRECTION.BEFORE,
+    QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT.TODAY,
+  ),
+];
+
+/**
+ * Generate pure date functions based on the DatePickerOption.
+ */
+const generatePureDateFunction = (
+  graph: PureModel,
+  datePickerOption: DatePickerOption,
+): SimpleFunctionExpression => {
+  const multiplicityOne = graph.getTypicalMultiplicity(
+    TYPICAL_MULTIPLICITY_TYPE.ONE,
+  );
+  const strictDate = graph.getPrimitiveType(PRIMITIVE_TYPE.STRICTDATE);
+  const date = graph.getPrimitiveType(PRIMITIVE_TYPE.DATE);
+  const dateTime = graph.getPrimitiveType(PRIMITIVE_TYPE.DATETIME);
+  switch (datePickerOption.value) {
```

Comment:
Here I think we can use the `instanceof` check so we don't end up with the `default` case and have to throw error.